File: scratch/update_json.py
import csv, json, re, sys
from pathlib import Path

ROOT = Path(__file__).parent.parent
CSV_PATH = ROOT / "EtsyListingsDownload.csv"
JSON_OUT = ROOT / "data" / "etsy_products.json"

DISCOUNT_RATE = 0.30
DEFAULT_RATING = 4.7
DEFAULT_REVIEW_COUNT = 285
DEFAULT_DIFFICULTY = "Easy"
DEFAULT_TIME = "8 hours"
DEFAULT_PAGES = 15

CATEGORY_MAP = [
    (["bar cart", "charcuterie", "serving cart", "beverage cart", "coffee cart",
      "mobile bar", "food cart", "vendor cart", "catering"], "Kitchen"),
    (["outdoor kitchen", "grill station", "bbq", "grill table"], "Kitchen"),
    (["farmstand", "farm stand", "farm cart", "flower cart", "produce", "roadside",
      "flower stand", "planter", "garden", "plant stand"], "Garden"),
    (["workbench", "workshop", "garage", "shop table", "tool bench"], "Workshop"),
    (["bed frame", "bunk bed", "loft bed", "kids", "playhouse", "treehouse"], "Bedroom"),
    (["digital", "digital download", "svg", "pdf", "plans"], "Digital"),
]

DIFFICULTY_HINTS = [
    (["folding", "foldable", "collapsible", "beginner", "simple"], "Easy"),
    (["rotating", "lift top", "lift-top", "ping pong", "outdoor kitchen", "modular"], "Hard"),
]

def slugify(text):
    text = text.lower().strip()
    text = re.sub(r"[^\w\s-]", "", text)
    text = re.sub(r"[\s_]+", "-", text)
    text = re.sub(r"-+", "-", text)
    return text[:120].strip("-")

def guess_category(title, desc=""):
    combined = (title + " " + desc[:200]).lower()
    for keywords, cat in CATEGORY_MAP:
        if any(kw in combined for kw in keywords):
            return cat
    return "Furniture"

def guess_difficulty(title, desc=""):
    combined = (title + " " + desc[:200]).lower()
    for keywords, diff in DIFFICULTY_HINTS:
        if any(kw in combined for kw in keywords):
            return diff
    return DEFAULT_DIFFICULTY

def build_image_url(url, size="fullxfull"):
    if not url: return ""
    return re.sub(r"il_[^./]+\.", f"il_{size}.", url)

def collect_images(row):
    full_images, thumb_images = [], []
    for i in range(1, 11):
        url = row.get(f"IMAGE{i}", "").strip()
        if url:
            full_images.append(build_image_url(url, "fullxfull"))
            thumb_images.append(build_image_url(url, "794xN"))
    return full_images, thumb_images

def parse_price(s):
    try: return float(re.sub(r"[^\d.]", "", s))
    except: return 0.0

def parse_tags(s):
    if not s: return []
    parts = [t.strip() for t in s.split(",") if t.strip()]
    return [re.sub(r"\s+", "_", t) for t in parts]

def load_exact_url_map(json_path):
    if not json_path.exists(): return {}
    try:
        data = json.loads(json_path.read_text(encoding="utf-8"))
        result = {}
        for p in data:
            name = p.get("name", "").strip()
            url = p.get("etsy_url", "")
            if name and url:
                result[name] = url
        return result
    except: return {}

def main():
    if not CSV_PATH.exists():
        print("ERROR: CSV not found")
        return

    url_map = load_exact_url_map(JSON_OUT)

    products = []
    idx = 1
    
    # User-provided NEW URLs map
    NEW_URLS = {
        "Tiered Planter Box Plans | 2-Level Raised Garden Bed Woodworking Blueprint (PDF Download)": 
            "https://www.etsy.com/listing/4493928541/tiered-planter-box-plans-2-level-raised",
        "Tiered Plant Stand Plans | DIY Outdoor Wooden Garden Shelf (PDF Download)":
            "https://www.etsy.com/listing/4493892575/tiered-plant-stand-plans-diy-outdoor",
    }
    
    # We will read etsy_products.json to preserve categories if we prefer, but for now we follow guess_category and update
    try:
        old_data = json.loads(JSON_OUT.read_text(encoding="utf-8"))
        cat_map_old = {p['name']: p.get('category') for p in old_data}
    except:
        cat_map_old = {}

    with open(CSV_PATH, encoding="utf-8-sig", errors="replace", newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            title = row.get("TITLE", "").strip()
            if not title: continue
            desc = row.get("DESCRIPTION", "").strip()
            raw_price = parse_price(row.get("PRICE", "0"))
            if raw_price <= 0: raw_price = 9.99

            sale_price = round(raw_price * DISCOUNT_RATE, 2)
            full_images, thumb_images = collect_images(row)
            tags = parse_tags(row.get("TAGS", ""))
            
            category = cat_map_old.get(title) # preserve old categories
            if not category:
                category = guess_category(title, desc)
                
            difficulty = guess_difficulty(title, desc)
            
            # Identify URL
            etsy_url = NEW_URLS.get(title)
            if not etsy_url:
                etsy_url = url_map.get(title, "")
                
            slug = slugify(title)
            short_desc = re.sub(r"\s+", " ", desc[:220]).strip()
            
            # Stock updates: reading from quantity if available, or setting a default stock status
            stock_quantity = int(row.get("QUANTITY", "0"))
            in_stock = stock_quantity > 0

            product = {
                "id": f"etsy-{idx}",
                "slug": slug,
                "name": title,
                "category": category,
                "price": sale_price,
                "originalPrice": raw_price,
                "rating": DEFAULT_RATING,
                "reviewCount": DEFAULT_REVIEW_COUNT,
                "difficulty": difficulty,
                "estimatedTime": DEFAULT_TIME,
                "pages": DEFAULT_PAGES,
                "description": short_desc,
                "longDescription": desc,
                "features": tags[:5],
                "tags": tags,
                "materials": ["Digital Download", "PDF File", "Instant Download"],
                "image": full_images[0] if full_images else "",
                "thumbnail": thumb_images[0] if thumb_images else "",
                "images": full_images,
                "imagesThumbnails": thumb_images,
                "etsy_url": etsy_url,
                "bestseller": False,
                "inStock": in_stock,
                "stockQuantity": stock_quantity
            }
            products.append(product)
            idx += 1

    JSON_OUT.write_text(json.dumps(products, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"Updated {len(products)} products! Two new URLs handled correctly. Stock and prices updated directly from CSV.")
    
if __name__ == "__main__":
    main()

File: scratch/update_json.js
const fs = require('fs');
const path = require('path');

const ROOT = __dirname;
const CSV_PATH = path.join(ROOT, '..', 'EtsyListingsDownload.csv');
const JSON_OUT = path.join(ROOT, '..', 'data', 'etsy_products.json');

const DISCOUNT_RATE = 0.30;
const DEFAULT_RATING = 4.7;
const DEFAULT_REVIEW_COUNT = 285;
const DEFAULT_DIFFICULTY = "Easy";
const DEFAULT_TIME = "8 hours";
const DEFAULT_PAGES = 15;

const CATEGORY_MAP = [
    [["bar cart", "charcuterie", "serving cart", "beverage cart", "coffee cart", "mobile bar", "food cart", "vendor cart", "catering"], "Kitchen"],
    [["outdoor kitchen", "grill station", "bbq", "grill table"], "Kitchen"],
    [["farmstand", "farm stand", "farm cart", "flower cart", "produce", "roadside", "flower stand", "planter", "garden", "plant stand"], "Garden"],
    [["workbench", "workshop", "garage", "shop table", "tool bench"], "Workshop"],
    [["bed frame", "bunk bed", "loft bed", "kids", "playhouse", "treehouse"], "Bedroom"],
    [["digital", "digital download", "svg", "pdf", "plans"], "Digital"],
];

const DIFFICULTY_HINTS = [
    [["folding", "foldable", "collapsible", "beginner", "simple"], "Easy"],
    [["rotating", "lift top", "lift-top", "ping pong", "outdoor kitchen", "modular"], "Hard"],
];

function slugify(text) {
    let t = text.toLowerCase().trim();
    t = t.replace(/[^\w\s-]/g, "");
    t = t.replace(/[\s_]+/g, "-");
    t = t.replace(/-+/g, "-");
    t = t.substring(0, 120);
    return t.replace(/^-+|-+$/g, "");
}

function guessCategory(title, desc = "") {
    const combined = (title + " " + desc.substring(0, 200)).toLowerCase();
    for (const [keywords, cat] of CATEGORY_MAP) {
        if (keywords.some(kw => combined.includes(kw))) {
            return cat;
        }
    }
    return "Furniture";
}

function guessDifficulty(title, desc = "") {
    const combined = (title + " " + desc.substring(0, 200)).toLowerCase();
    for (const [keywords, diff] of DIFFICULTY_HINTS) {
        if (keywords.some(kw => combined.includes(kw))) {
            return diff;
        }
    }
    return DEFAULT_DIFFICULTY;
}

function buildImageUrl(url, size = "fullxfull") {
    if (!url) return "";
    return url.replace(/il_[^.]+\./, `il_${size}.`);
}

function parsePrice(s) {
    const val = parseFloat(s.replace(/[^\d.]/g, ""));
    return isNaN(val) ? 0.0 : val;
}

function parseTags(s) {
    if (!s) return [];
    return s.split(",").map(t => t.trim().replace(/\s+/g, "_")).filter(t => t);
}

const NEW_URLS = {
    "Tiered Planter Box Plans | 2-Level Raised Garden Bed Woodworking Blueprint (PDF Download)": "https://www.etsy.com/listing/4493928541/tiered-planter-box-plans-2-level-raised",
    "Tiered Plant Stand Plans | DIY Outdoor Wooden Garden Shelf (PDF Download)": "https://www.etsy.com/listing/4493892575/tiered-plant-stand-plans-diy-outdoor",
};

function main() {
    if (!fs.existsSync(CSV_PATH)) {
        console.error("ERROR: CSV not found");
        return;
    }

    let catMapOld = {};
    let urlMapOld = {};
    try {
        const oldData = JSON.parse(fs.readFileSync(JSON_OUT, 'utf-8'));
        for (const p of oldData) {
            if (p.name && p.category) catMapOld[p.name.trim()] = p.category;
            if (p.name && p.etsy_url) urlMapOld[p.name.trim()] = p.etsy_url;
        }
    } catch (e) {
        // ignore
    }

    const csvData = fs.readFileSync(CSV_PATH, 'utf-8');
    const lines = [];
    let insideQuote = false;
    let currentLine = "";
    for (let char of csvData) {
        if (char === '"') insideQuote = !insideQuote;
        if (char === '\n' && !insideQuote) {
            lines.push(currentLine);
            currentLine = "";
        } else {
            currentLine += char;
        }
    }
    if (currentLine) lines.push(currentLine);

    if (lines.length === 0) return;
    
    // Parse headers
    function parseCSVLine(line) {
        const row = [];
        let cur = "";
        let inQ = false;
        for (let i = 0; i < line.length; i++) {
            const ch = line[i];
            if (ch === '"') {
                if (inQ && line[i+1] === '"') {
                    cur += '"';
                    i++;
                } else {
                    inQ = !inQ;
                }
            } else if (ch === ',' && !inQ) {
                row.push(cur);
                cur = "";
            } else {
                cur += ch;
            }
        }
        row.push(cur);
        return row;
    }

    const unbommed = lines[0].replace(/^\uFEFF/, '');
    const headers = parseCSVLine(unbommed.trim());
    const headerIndex = {};
    headers.forEach((h, i) => headerIndex[h.trim()] = i);

    const products = [];
    let idx = 1;

    for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;
        const rowArr = parseCSVLine(line);
        const row = {};
        for(let j=0; j<headers.length; j++){
            row[headers[j].trim()] = rowArr[j] || "";
        }

        const title = (row["TITLE"] || "").trim();
        if (!title) continue;

        const desc = (row["DESCRIPTION"] || "").trim();
        let rawPrice = parsePrice(row["PRICE"] || "0");
        if (rawPrice <= 0) rawPrice = 9.99;

        const salePrice = Math.round(rawPrice * DISCOUNT_RATE * 100) / 100;
        
        const fullImages = [];
        const thumbImages = [];
        for (let j = 1; j <= 10; j++) {
            let url = (row[`IMAGE${j}`] || "").trim();
            if (url) {
                fullImages.push(buildImageUrl(url, "fullxfull"));
                thumbImages.push(buildImageUrl(url, "794xN"));
            }
        }

        const tags = parseTags(row["TAGS"] || "");
        let category = catMapOld[title];
        if (!category) category = guessCategory(title, desc);
        const difficulty = guessDifficulty(title, desc);

        let etsyUrl = NEW_URLS[title] || urlMapOld[title] || "";

        const slug = slugify(title);
        const shortDesc = desc.replace(/\s+/g, " ").substring(0, 220).trim();

        const quantityText = row["QUANTITY"] || "0";
        const stockQuantity = parseInt(quantityText, 10);
        const inStock = stockQuantity > 0;

        products.push({
            id: `etsy-${idx}`,
            slug: slug,
            name: title,
            category: category,
            price: salePrice,
            originalPrice: rawPrice,
            rating: DEFAULT_RATING,
            reviewCount: DEFAULT_REVIEW_COUNT,
            difficulty: difficulty,
            estimatedTime: DEFAULT_TIME,
            pages: DEFAULT_PAGES,
            description: shortDesc,
            longDescription: desc,
            features: tags.slice(0, 5),
            tags: tags,
            materials: ["Digital Download", "PDF File", "Instant Download"],
            image: fullImages[0] || "",
            thumbnail: thumbImages[0] || "",
            images: fullImages,
            imagesThumbnails: thumbImages,
            etsy_url: etsyUrl,
            bestseller: false,
            inStock: inStock,
            stockQuantity: isNaN(stockQuantity) ? 0 : stockQuantity
        });
        idx++;
    }

    fs.writeFileSync(JSON_OUT, JSON.stringify(products, null, 2), 'utf-8');
    console.log(`Updated ${products.length} products!`);
}

main();
